sdk/nodejs/cmd/pulumi-language-nodejs/main.go

```diff
@@ -362,84 +364,192 @@ func getPluginVersion(info packageJSON) (string, error) {
+		},
+	}, tracingSpan)
+	if err != nil {
+		return nil, err
 	}
 
-	if logging.V(5) {
-		commandStr := strings.Join(args, " ")
-		logging.V(5).Infoln("Language host launching process: ", host.nodeBin, commandStr)
+	// Create the pipes we'll use to communicate synchronously with the nodejs process. Once we're
+	// done using the pipes clean them up so we don't leave anything around in the user file system.
+	pipes, pipesDone, err := createAndServePipes(ctx, target)
+	if err != nil {
+		return nil, err
 	}
 
-	// Now simply spawn a process to execute the requested program, wiring up stdout/stderr directly.
-	var errResult string
-	cmd := exec.Command(host.nodeBin, args...)
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	cmd.Env = env
-	if err := cmd.Run(); err != nil {
-		// NodeJS stdout is complicated enough that we should explicitly flush stdout and stderr here. NodeJS does
-		// process writes using console.out and console.err synchronously, but it does not process writes using
-		// `process.stdout.write` or `process.stderr.write` synchronously, and it is possible that there exist unflushed
-		// writes on those file descriptors at the time that the Node process exits.
-		//
-		// Because of this, we explicitly flush stdout and stderr so that we are absolutely sure that we capture any
-		// error messages in the engine.
-		contract.IgnoreError(os.Stdout.Sync())
-		contract.IgnoreError(os.Stderr.Sync())
-		if exiterr, ok := err.(*exec.ExitError); ok {
-			// If the program ran, but exited with a non-zero error code.  This will happen often,
-			// since user errors will trigger this.  So, the error message should look as nice as
-			// possible.
-			if status, stok := exiterr.Sys().(syscall.WaitStatus); stok {
-				// Check if we got special exit code that means "we already gave the user an
-				// actionable message". In that case, we can simply bail out and terminate `pulumi`
-				// without showing any more messages.
-				if status.ExitStatus() == nodeJSProcessExitedAfterShowingUserActionableMessage {
-					return &pulumirpc.RunResponse{Error: "", Bail: true}, nil
-				}
+	// Channel producing the final response we want to issue to our caller. Will get the result of
+	// the actual nodejs process we launch, or any results caused by errors in our server/pipes.
+	responseChannel := make(chan *pulumirpc.RunResponse)
+	defer close(responseChannel)
 
-				err = errors.Errorf("Program exited with non-zero exit code: %d", status.ExitStatus())
+	// Forward any rpc server or pipe errors to our output channel.
+	go func() {
+		err := <-serverDone
+		if err != nil {
+			responseChannel <- &pulumirpc.RunResponse{Error: err.Error()}
+		}
+	}()
+	go func() {
+		err := <-pipesDone
+		if err != nil {
+			responseChannel <- &pulumirpc.RunResponse{Error: err.Error()}
+		}
+	}()
+
+	// now, launch the nodejs process and actually run the user code in it.
+	go host.execNodejs(responseChannel, req, fmt.Sprintf("127.0.0.1:%d", port), pipes.directory())
```

Comment:
@pgavlin this was a place i couldn't figure out how to make synchronous.  i.e. it feels like it needs ot be a goroutine.  Specifically, if this is synchronous then i don't really understand how we can deal with its result *and* also deal with the results of concurrent work being done by rpcutil.Serve or createAndServePipes.  If any of these three might report their own error (concurrently), then what is the right way to observe and report those back properly?  Thanks!

Reply:
It's fine to leave this as-is. We can always clean it up later.

Reply:
thanks.